File: python/pose.py
import numpy as np
from scipy.stats import multivariate_normal
import math


def matM(nu, omega, time, stds):
    return np.diag([stds["nn"]**2*abs(nu)/time + stds["no"]**2*abs(omega)/time, 
                    stds["on"]**2*abs(nu)/time + stds["oo"]**2*abs(omega)/time])
    
def matA(nu, omega, time, theta):
    st, ct = math.sin(theta), math.cos(theta)
    stw, ctw = math.sin(theta + omega*time), math.cos(theta + omega*time)
    return np.array([[(stw - st)/omega,    -nu/(omega**2)*(stw - st) + nu/omega*time*ctw],
                    [(-ctw + ct)/omega, -nu/(omega**2)*(-ctw + ct) + nu/omega*time*stw],
                    [0,                                time]] )

def matF(nu, omega, time, theta):
    F = np.diag([1.0, 1.0, 1.0])
    F[0, 2] = nu / omega * (math.cos(theta + omega * time) - math.cos(theta))
    F[1, 2] = nu / omega * (math.sin(theta + omega * time) - math.sin(theta))
    return F

def matH(pose, poseo): ###kf4funcs
    mx, my = poseo
    mux, muy = pose
    q = (mux - mx)**2 + (muy - my)**2
    return np.array([[(mux - mx)/np.sqrt(q), (muy - my)/np.sqrt(q), 0.0],  [(my - muy)/q, (mux - mx)/q, -1.0]])

def matQ(distance_dev, direction_dev):
    return np.diag(np.array([distance_dev**2, direction_dev**2]))


class KalmanFilter_Pose():
    def __init__(self, initial_state_mean, acc, system_cov={"nn":0.19, "no":0.001, "on":0.13, "oo":0.2},):
        self.belief = multivariate_normal(mean=initial_state_mean, cov=np.diag([1e-10, 1e-10, 1e-10])) 
        self.system_cov = np.array(system_cov) #システムノイズの共分散行列
        self.distance_dev_rate=0.14
        self.direction_dev=0.05
        # self.initial_state_mean = np.array(initial_state_mean) #状態の初期値の平均ベクトル
        # self.initial_state_cov = np.array(initial_state_cov) #状態の初期値の共分散行列
        self.acc_change = np.zeros(acc.size)

        
    def filter_predict(self, nu, omega, time):
        M = matM(nu, omega, time, self.system_cov)
        A = matA(nu, omega, time, self.belief.mean[2])
        F = matF(nu, omega, time, self.belief.mean[2])
        predicted_state_mean = F @ self.belief.mean
        predicted_state_cov = (F @ self.belief.cov @ (F.T) + A @ M @ (A.T))

        return (predicted_state_mean, predicted_state_cov)

    def filter_update(self, predicted_state_mean, predicted_state_cov, pose):
        H = matH(pose[:1], self.belief.mean[:1])
        estimated_z = self.observation_function(self.belief.mean, predicted_state_mean)
        z = np.array(np.hypot([pose[:1] - self.belief.mean[:1]]) )
        Q = matQ(estimated_z[0]*self.distance_dev_rate, self.direction_dev)
        kalman_gain = (predicted_state_cov @ (H.T) @ np.linalg.inv(H @ predicted_state_cov @ (H.T)+ Q) )
        filtered_state_mean = (predicted_state_mean + kalman_gain @ (z - estimated_z) )
        filtered_state_cov = (predicted_state_cov - (kalman_gain @ H @ predicted_state_cov) )
        self.acc_change = filtered_state_mean - self.belief.mean
        self.belief.mean = filtered_state_mean
        self.belief.cov = filtered_state_cov

        return (filtered_state_mean, filtered_state_cov)
        
    def observation_function(poseo, pose_estimate):
        diff = pose_estimate - poseo[0:2]
        phi = math.atan2(diff[1], diff[0]) - poseo[2]
        # while phi >= np.pi: phi -= 2*np.pi
        # while phi < -np.pi: phi += 2*np.pi
        return np.array( [np.hypot(*diff), phi ] ).T
    

    def pose_estimate(self, x, y, theta, time):
        pose = (x, y, theta)
        nu = (np.hypot(pose[:1] - self.belief.mean[:1]))/time
        omega = theta / time
        # フィルタリング分布取得
        predicted_state_mean, predicted_state_cov = self.filter_predict(nu, omega, time)
        filtered_state_mean, filtered_state_cov = self.filter_update(predicted_state_mean, predicted_state_cov, pose)

        return filtered_state_mean

File: src/kf/pose.rs
use nalgebra as na;
use std::f32::consts::PI;

pub struct Stds {
    pub nn: f32,
    pub no: f32,
    pub on: f32,
    pub oo: f32,
}

pub fn mat_m(nu: f32, omega: f32, time: f32, stds: &Stds) -> na::Matrix2<f32> {
    let mat = na::Matrix2::new(
        stds.nn.powf(2.)*nu.abs()/time + stds.no.powf(2.)*omega.abs()/time,   0.,
        0.  , stds.on.powf(2.)*nu.abs()/time + stds.oo.powf(2.)*omega.abs()/time);
    mat
}

fn mat_a(nu: f32, omega: f32, time: f32, theta: f32) -> na::Matrix3x2<f32> {
    let st = theta.sin();
    let ct = theta.cos();
    let stw = (theta + omega * time).sin();
    let ctw = (theta + omega * time).cos();
    let mat = na::Matrix3x2::new(
        (-stw + st)/omega, (-nu/(omega.powf(2.0))*(stw - st))+(nu/omega*time*ctw), 
        (-ctw + ct)/omega, (-nu/(omega.powf(2.0))*(-ctw + ct))+(nu/omega*time*stw),
        0.               , time);
    mat
}

fn mat_f(nu: f32, omega: f32, time: f32, theta: f32) -> na::Matrix3<f32> {
    let mut mat = na::Matrix3::identity();
    mat[(0, 2)] = (nu/omega)*(theta + omega*time).cos() - (nu/omega)*theta.cos();
    mat[(1, 2)] = (nu/omega)*(theta + omega*time).sin() - (nu/omega)*theta.sin();
    mat
}

fn mat_h(pose: &na::Vector3<f32>, poseo: &na::Vector3<f32>) -> na::Matrix2x3<f32> {
    let mx = poseo[0];
    let my = poseo[1];
    let mux = pose[0];
    let muy = pose[1];
    let q = (mux - mx).powf(2.0) + (muy - my).powf(2.0);
    let mat = na::Matrix2x3::new(
        (mux - mx)/q.sqrt(), (muy - my)/q.sqrt(), 0.,
        (my - muy)/q       , (mux - mx)/q       , -1.0);
    mat
}

fn mat_q(distance_dev: f32, direction_dev: f32) -> na::Matrix2<f32> {
    let mat = na::Matrix2::new(
        distance_dev.powf(2.0), 0.,
        0., direction_dev.powf(2.0));
    mat
}

pub struct KalmanFilterPose <'a> {
    belief: Belief,
    system_cov: & 'a Stds,
    distance_dev_rate: f32,
    direction_dev: f32,
    pose_change: na::Vector3<f32>,
}

struct Belief {
    mean: na::Vector3<f32>,
    cov: na::Matrix3<f32>,
}

impl <'a> KalmanFilterPose <'a> {
    pub fn new( initial_state_mean: &na::Vector3<f32>, initial_cov: f32) -> Self {
        let system_cov = &Stds {nn:0.19, no:0.001, oo:0.13, on:0.2,};
        let belief = Belief { mean: *initial_state_mean, cov: initial_cov * na::Matrix3::identity()};
        Self{ belief, system_cov, distance_dev_rate: 0.14, direction_dev: 0.05, pose_change: na::Vector3::zeros() }
    }

    pub fn filter_predict(&self, nu: f32, omega: f32, time: f32) -> (na::Vector3<f32>, na::Matrix3<f32>){
        let m = mat_m(nu, omega, time, self.system_cov);
        let a = mat_a(nu, omega, time, self.belief.mean[2]);
        let f = mat_f(nu, omega, time, self.belief.mean[2]);
        let predicted_state_mean = f * (&self.belief.mean);
        let predicted_state_cov = f * (&self.belief.cov) * (&f.transpose()) + a * (&m) * (&a.transpose());
        (predicted_state_mean, predicted_state_cov)
    }

    pub fn filter_update(& mut self, predicted_state_mean: &na::Vector3<f32>, predicted_state_cov: &na::Matrix3<f32>, pose: &na::Vector3<f32>)-> (na::Vector3<f32>, na::Matrix3<f32>){
        let h = mat_h(&pose, &self.belief.mean);
        let estimated_z = self.observation_function(&self.belief.mean, &predicted_state_mean);
        // let z = na::Vector2::new(pose - self.belief.mean);
        let z = self.observation_function(&pose, &self.belief.mean);
        let q = mat_q(estimated_z[0]*self.distance_dev_rate, self.direction_dev);
        let kalman_gain = predicted_state_cov * (h.transpose()) * (h * predicted_state_cov * h.transpose() + q).try_inverse().unwrap();
        // let kalman_gain = predicted_state_cov * (&h.transpose()) * (&h * (&predicted_state_cov) * (&h.transpose()) + q).try_inverse().unwrap();
        let filtered_state_mean = predicted_state_mean + kalman_gain * (z - estimated_z);
        let filtered_state_cov = predicted_state_cov - ( kalman_gain * h * predicted_state_cov);
        self.pose_change = filtered_state_mean - self.belief.mean;
        self.belief.mean = filtered_state_mean;
        self.belief.cov = filtered_state_cov;
        (filtered_state_mean, filtered_state_cov)
    }

    pub fn observation_function(&self, poseo: &na::Vector3<f32>, pose_estimate: &na::Vector3<f32>) -> na::Matrix2x1<f32> {
        let diff = pose_estimate - poseo;
        let mut phi = diff[1].atan2(diff[0]) - poseo[2];
        while phi >= PI { phi -= 2. * PI; }
        while phi < -PI { phi += 2. * PI; }
        let mat = na::Matrix2x1::new(diff[0].hypot(diff[1]), phi);
        mat
    }

    // pub fn pose_estimate(& mut self, x: f32, y: f32, theta: f32, time: f32) -> na::Vector3<f32> {
    //     let pose = na::Vector3::new(x, y, theta);
    //     let nu = (pose[1] - self.belief.mean[1]).hypot(pose[0] - self.belief.mean[0])/time;
    //     let omega = theta / time;
    //     // フィルタリング分布取得
    //     let predicted_state = self.filter_predict(nu, omega, time);
    //     let filtered_state = self.filter_update(&predicted_state.0, &predicted_state.1, &pose);

    //     return filtered_state.0
    // }
}